fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+            return 1
+        return multiprocessing.cpu_count()
+
+    if not isinstance(num_workers, int) or num_workers < 1:
+        raise ValueError(
+            "The `num_workers` argument must be a positive integer or `None` "
+            "found %s" % str(type(num_workers))
+        )
+    return num_workers
+
+
+def _build_tasks_list(
+    urls, clip_segments, download_dir, ext, buffer_seconds=5
+):
+    if isinstance(urls, str):
+        urls = [urls]
+    if isinstance(clip_segments, tuple):
+        clip_segments = [clip_segments]
+
+    if isinstance(urls, list):
+        if download_dir is None:
+            raise ValueError(
+                "When `urls` is a list, `download_dir` is required but was found to be `None`."
+            )
+        urls = {url: None for url in urls}
+
+        etau.ensure_dir(download_dir)
+
+    num_videos = len(urls)
+
+    if clip_segments is None:
+        clip_segments = [None] * num_videos
+
+    if len(clip_segments) != num_videos:
+        raise ValueError(
+            "Found %d `clip_segments` and %d `urls`, but the lengths of these "
+            "iterables must match" % (len(clip_segments), num_videos)
+        )
+
+    ext_list = _parse_list_arg(ext, num_videos)
+    urls_list = _parse_list_arg(list(urls.keys()), num_videos)
+    paths_list = _parse_list_arg(list(urls.values()), num_videos)
+    download_dir_list = [download_dir] * num_videos
+    buffer_seconds_list = [buffer_seconds] * num_videos
+
+    return list(
+        zip(
+            urls_list,
+            paths_list,
+            clip_segments,
+            download_dir_list,
+            ext_list,
+            buffer_seconds_list,
+        )
+    )
+
+
+def _parse_list_arg(arg, list_len):
+    if isinstance(arg, list):
+        if len(arg) != list_len:
+            # If arg list is not the right length, ignore it and set all arg to
+            # `None`
+            arg_list = [None] * list_len
+        else:
+            arg_list = arg
+    else:
+        # If a non-list arg is given, repeat it for the length of the list
+        arg_list = [arg] * list_len
+
+    return arg_list
+
+
+def _single_thread_download(tasks, max_videos):
+    downloaded = []
+    errors = defaultdict(list)
+    with fou.ProgressBar(total=max_videos, iters_str="videos") as pb:
+        for task in tasks:
+            is_success, url, error_type = _do_download(task)
+            if is_success:
+                downloaded.append(url)
+                pb.update()
+                if len(downloaded) >= max_videos:
+                    return downloaded, errors
+            else:
+                errors[error_type].append(url)
+    return downloaded, errors
+
+
+def _multi_thread_download(tasks, max_videos, num_workers):
+    downloaded = []
+    errors = defaultdict(list)
+    with fou.ProgressBar(total=max_videos, iters_str="videos") as pb:
+        with multiprocessing.dummy.Pool(num_workers) as pool:
```

Comment:
`multiprocessing.dummy.Pool` uses threads, not processes.

Ideally, full video downloads would be done using a thread pool, since the bottleneck is network bandwidth, while clip downloads would be done using a process pool, since here the bottleneck is some combination of network and compute (`ffmpeg`) bandwidth.

Reply:
Processes could *always* be used on Linux/Mac, but the trouble is that they are sketchy on Windows.

So, the ideal compromise might be preferable to offer this functionality as separate methods:
- `download_youtube_videos()` --- thread pool
- `download_youtube_clips()` --- process pool

Reply:
It turned out to be pretty clean to switch between threads and processes without using two functions.